### ISSUE TITLE
error.c was lost while compiling the instrgen executable file

### ISSUE THREAD

**Issue by huanghantao** [freewilll/wcc#2]
```bash
make wcc
......
gcc -o instrgen instrgen.c instrrules.c instrutil.c utils.c memory.c longmap.c types.c scopes.c list.c set.c strmap.c
1 warning generated.
Undefined symbols for architecture x86_64:
  "_error", referenced from:
      _integer_promote_type in types-dd89eb.o
      _deref_pointer in types-dd89eb.o
      _composite_type in types-dd89eb.o
      _flatten_anonymous_structs in types-dd89eb.o
     (maybe you meant: _opt_warnings_are_errors)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [instrgen] Error 1
```

I tried to include the `error.c` file when compiling, but `error.c` relies on too many other source files

**Comment by freewilll** [comment]
Can you give me some details about your environment? I see you are using clang? I developed this with gcc and Ubuntu. What happens if you try gcc? I can help much at the moment since I won't be near a laptop for a couple of weeks due to holiday.

**Comment by huanghantao** [comment]
yes, I use clang:

```bash
Configured with: --prefix=/Library/Developer/CommandLineTools/usr --with-gxx-include-dir=/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/c++/4.2.1
Apple clang version 12.0.0 (clang-1200.0.32.28)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

**Comment by freewilll** [comment]
I just tried this on my mac and it compiles fine (albeit with a ton of warnings).

```
gcc -o instrgen instrgen.c instrrules.c instrutil.c utils.c memory.c longmap.c types.c scopes.c list.c set.c strmap.c error.c cpp.c strset.c lexer.c constexpr.c parser.c functions.c ir.c codegen.c longset.c instrsel.c graph.c ssa.c stack.c  -Wno-return-type -D BUILD_DIR='"/Users/wangenent/code/wcc/src"'
parser.c:1067:76: warning: too many arguments in call to 'parse_declaration_specifiers'
    BaseType *base_type_with_storage_class = parse_declaration_specifiers(0);
                                             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ^
parser.c:1751:93: warning: logical not is only applied to the left hand side of this comparison [-Wlogical-not-parentheses]
        if (!scalar->is_constant && !scalar->is_string_literal && !scalar->is_address_of && !scalar->type->type == TYPE_ARRAY)
                                                                                            ^                   ~~
parser.c:1751:93: note: add parentheses after the '!' to evaluate the comparison first
        if (!scalar->is_constant && !scalar->is_string_literal && !scalar->is_address_of && !scalar->type->type == TYPE_ARRAY)
                                                                                            ^
                                                                                             (                               )
parser.c:1751:93: note: add parentheses around left hand side expression to silence this warning
        if (!scalar->is_constant && !scalar->is_string_literal && !scalar->is_address_of && !scalar->type->type == TYPE_ARRAY)
                                                                                            ^
                                                                                            (                  )
parser.c:1751:113: warning: result of comparison of constant 'TYPE_ARRAY' (10) with boolean expression is always false [-Wtautological-constant-out-of-range-compare]
        if (!scalar->is_constant && !scalar->is_string_literal && !scalar->is_address_of && !scalar->type->type == TYPE_ARRAY)
                                                                                            ~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~
3 warnings generated.
codegen.c:716:26: warning: result of comparison of constant 128 with expression of type 'char' is always false [-Wtautological-constant-out-of-range-compare]
    else if (c < 32 || c >= 128) return fprintf_octal_char(f, c);
                       ~ ^  ~~~
1 warning generated.
instrsel.c:1374:37: warning: too many arguments in call to 'simplify_igraphs'
        simplify_igraphs(function, i);
        ~~~~~~~~~~~~~~~~            ^
1 warning generated.
./instrgen > instrrules-generated.c
...
```

Your command line looks a bit odd, it seems to be missing a bunch of things, starting with `error.c cpp.c ...`. See my version above. If I run your command line I get the same error. Can you paste the output of `make clean && make instrgen`?

**Comment by huanghantao** [comment]
I get the same error:

```bash
 ~/codeDir/cCode/wcc   master  make clean && make instrgen
cd tests && /Library/Developer/CommandLineTools/usr/bin/make clean
cd unit && /Library/Developer/CommandLineTools/usr/bin/make clean
cd integration && /Library/Developer/CommandLineTools/usr/bin/make clean
cd e2e && /Library/Developer/CommandLineTools/usr/bin/make clean
cd tools && /Library/Developer/CommandLineTools/usr/bin/make clean
gcc -o instrgen instrgen.c instrrules.c instrutil.c utils.c memory.c longmap.c types.c scopes.c list.c set.c strmap.c
instrrules.c:1433:1: warning: non-void function does not return a value [-Wreturn-type]
}
^
instrrules.c:1439:1: warning: non-void function does not return a value [-Wreturn-type]
}
^
2 warnings generated.
instrutil.c:25:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
}
^
instrutil.c:266:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
}
^
instrutil.c:483:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
}
^
3 warnings generated.
utils.c:84:1: warning: non-void function does not return a value [-Wreturn-type]
}
^
utils.c:106:1: warning: non-void function does not return a value [-Wreturn-type]
}
^
2 warnings generated.
types.c:462:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
}
^
types.c:504:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
}
^
types.c:1029:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
}
^
types.c:1060:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
}
^
types.c:1073:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
}
^
types.c:1088:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
}
^
6 warnings generated.
set.c:74:1: warning: non-void function does not return a value [-Wreturn-type]
}
^
1 warning generated.
Undefined symbols for architecture x86_64:
  "_error", referenced from:
      _integer_promote_type in types-b9849e.o
      _deref_pointer in types-b9849e.o
      _composite_type in types-b9849e.o
      _flatten_anonymous_structs in types-b9849e.o
     (maybe you meant: _opt_warnings_are_errors)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [instrgen] Error 1
```

**Comment by huanghantao** [comment]
https://github.com/freewilll/wcc/blob/0a6b7ba3bfc9eb00cdaff9dd3485bfa133b58c86/Makefile#L47

It seems that the source file does not contain `error.c`

**Comment by freewilll** [comment]
I developed wcc entirely on linux and even attempt to make it work on other platforms such as MacOS. I fixed this issue. However you won't get very far on MacOS. Even if wcc were to produce an assembly file, it'll be for x86_64 linux, not MacOS.

Trying to compile something results in:

```
$ ./wcc main.c -c
main.c:1: error: Unable to find stdio.h in any include paths
```
This is because `BUILTIN_INCLUDE_PATHS` in cpp.c has hardcoded linux paths in it.

If I fix that by hardcoding `"/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/"`,
I don't get much further:
```
$ ./wcc main.c -c
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/cdefs.h:81: warning: warning "Unsupported compiler detected"
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/cdefs.h:458: error: Expected identifier
```

This is probably a bug.

I'll close this issue since, technically, the bug has been fixed. However, if you want wcc to work on MacOS, it will require some further development and testing.
